src/Tutorial.ts

```diff
@@ -0,0 +1,438 @@
+            }
+        })
+    }
+
+    //cooldown function that prevents too many actions that would cause the tutorial steps to go out of whack
+    startCooldown = () : void => {
+        cooldown = true
+        setTimeout(function(){
+            cooldown = false
+        }, 500)
+    }
+
+    initiateTutStep = (direction:string) :void => {
+        const eagle = Eagle.getInstance()
+
+        var tutStep = activeTut.getTutorialSteps()[activeTutCurrentStep-1]
+
+        //if there is a preFunction set, then we execute it here
+        var preFunction
+
+        if(direction === 'next'){
+            preFunction = tutStep.getPreFunct()
+        }else if(direction === 'prev'){
+            preFunction = tutStep.getBackPreFunct()
+        }
+        
+        if(preFunction != null){
+            preFunction(eagle)
+        }
+        
+        //we always pass through the wait funciton, it is decided there if we actully wait or not
+        this.initiateWaitForElement(tutStep.getWaitType())
+    }
+
+    initiateWaitForElement = (waitType:Wait.Type) :void => {
+        if(waitType===Wait.Type.None){
+            this.pickStepType(activeTut.getTutorialSteps()[activeTutCurrentStep-1],null)
+        }else{
+            //we set a two second timer, the wait will check every .1 seconds for two seconds at which point it is timed out and we abort the tut
+            waitForElementTimer = setInterval(function(){activeTut.waitForElement(waitType)}, 100);  
+            setTimeout(function(){
+                if(waitForElementTimer != null){
+                    clearTimeout(waitForElementTimer);
+                    waitForElementTimer = null;
+                    console.warn('waiting for next tutorial step element timed out')
+                }
+            }, 2000)
+        }
+    }
+
+    waitForElement = (waitType:Wait.Type) :void => {
+        var tutStep = activeTut.getTutorialSteps()[activeTutCurrentStep-1]
+        var elementAvailable = false
+        var selectorElement = tutStep.getSelector()
+        var alternateHighlightSelector = null
+
+        if(waitType === Wait.Type.Modal){
+            //in  case of a modal we make sure the selector is for the modal, we then check if it has the class 'show'
+            //we also pass this modal selector to the highlighting function, so whole modal is highlighted, 
+            //but the arrow still points at a specific object in the modal
+            if(!selectorElement().hasClass('modal')){
+                selectorElement = selectorElement().closest('.modal')
+                alternateHighlightSelector = selectorElement
+            }else{
+                selectorElement = selectorElement()
+            }
+            elementAvailable = selectorElement.hasClass('show')
+        }else if (waitType === Wait.Type.Element){
+            if(selectorElement().length){
+                elementAvailable = true
+            }else{
+                return
```

Comment:
I think this case should add a warning/error to the console.

If were waiting for an element, and zero elements were found using the selector, then we should report an error.